### PR TITLE
fixes blank page bug when adding large elements on new pages

### PR DESCRIPTION
**Pull request by semaeostomea**
When adding a large element (a cell, image, etc.) to a newly added page that would trigger the page break, a blank page would be added before adding a page with the actual element. 
I added a new variable `.page_body_has_content` that is set to `True` after any kind of content is added to the page (excluding the header), which is checked against in `.will_page_break`. If the page is empty, no page break is triggered.

This bug also affected the [unit_int.pdf](https://github.com/PyFPDF/fpdf2/blob/master/test/layout/unit_int.pdf) because of the extra padding of the cell triggering the page break, so I had to generate a new one.

This extra padding (like with any element that is larger than the effective page height) was simply cut off on the second page, so preventing the page break on an empty page does not cause any differences; only the blank page is omitted.

**Checklist**:

- [x] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged 

- [x] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/PyFPDF/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by semaeostomea** [comment]
I'm not sure why the check is throwing a certificate error on `./tutorial/tuto1.pdf` now, even before even running the tests. My commit didn't change that one 🤔

**Comment by gmischler** [comment]
> I'm not sure why the check is throwing an error now, even before running the tests.

It threw an error when trying to download one of the PDF checkers.
At about the same time you were unable to reach the site.
Both of those point to Github experiencing a temporary network outage.

I really hope that wasn't your fault! :wink:

**Comment by Lucas-C** [comment]
I will investigate & fix this in #530

**Comment by Lucas-C** [comment]
#530 ha been solved

Could you please rebase your branch on `mater` @semaeostomea?
That should solve the failing GitHub Action pipeline 😊

**Comment by semaeostomea** [comment]
> #530 has been solved
> 
> Could you please rebase your branch on `master` @semaeostomea? That should solve the failing GitHub Action pipeline 😊

Done, thank you 😊

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -3155,6 +3172,7 @@ def will_page_break(self, height):
         return (
             self.y + height > self.page_break_trigger
             and not self.in_footer
+            and self.page_body_has_content
```
Comment (Lucas-C) [inline fpdf/fpdf.py:3175]: I'm not a big fan of introducing a new attribute on the `FPDF` class...
I already has too many properties.
I have tested the following code and I think it should behave similarly:
```python
        return (
            self.y > self.t_margin  # ensure that there is already some content on the page
            and self.y + height > self.page_break_trigger
            and not self.in_footer
            and self.accept_page_break
        )
```

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Thank you for submitting this PR!

One thing is missing: some unit tests covering your case.

I suggest to add the following ones in `test_full_page_image.py`:
```python
def test_full_pdf_width_image(tmp_path):  # issue-528
    pdf = fpdf.FPDF()
    pdf.set_margin(0)
    pdf.add_page()
    pdf.image(HERE / "png_images/51a4d21670dc8dfa8ffc9e54afd62f5f.png", w=pdf.epw)
    assert_pdf_equal(pdf, HERE / "full_pdf_width_image.pdf", tmp_path)


def test_full_pdf_height_image(tmp_path):  # issue-528
    pdf = fpdf.FPDF()
    pdf.set_margin(0)
    pdf.add_page()
    pdf.image(HERE / "png_images/51a4d21670dc8dfa8ffc9e54afd62f5f.png", h=pdf.eph)
    assert_pdf_equal(pdf, HERE / "full_pdf_height_image.pdf", tmp_path)
```

I have tested them and they should cover the edge case you have spotted (especially `test_full_pdf_width_image`)

File: fpdf/fpdf.py (hunk 1 above)
Comment (semaeostomea) [reply fpdf/fpdf.py:3175]: > I'm not a big fan of introducing a new attribute on the `FPDF` class... It already has too many properties. I have tested the following code and I think it should behave similarly:
> 
> ```python
>         return (
>             self.y > self.t_margin  # ensure that there is already some content on the page
>             and self.y + height > self.page_break_trigger
>             and not self.in_footer
>             and self.accept_page_break
>         )
> ```

The problem with this is that it wouldn't work when y is set before inserting the element (which I for example often do, but that might just be me). I couldn't see a way around adding a new attribute that would cover all use cases

Comment (Lucas-C) [reply fpdf/fpdf.py:3175]: > The problem with this is that it wouldn't work when y is set before inserting the element (which I for example often do, but that might just be me). I couldn't see a way around adding a new attribute that would cover all use cases

OK I see.

Then maybe the heuristic could be this:
```python
        # The magic number 11 just below is len(bytearray(b'2 J\n0.57 w\n')),
        # the value of self.pages[self.page]["content"] just after _beginpage() has been called:
        page_has_content = self.y > self.t_margin and self.page > 0 and len(self.pages[self.page]["content"]) > 11
        return (
            page_has_content
            and self.y + height > self.page_break_trigger
            and not self.in_footer
            and self.accept_page_break
        )
```

What do you think @semaeostomea & @gmischler?

Comment (gmischler) [reply fpdf/fpdf.py:3175]: Independently of whether there's already other content on a page, `self.y <= self.t_margin` might still be a valid test for not triggering an auto page break. After all, if an element won't fit on the page under those circumstances, then it won't fit any better on the next one either.

The auto-page-break feature was designed for flowing text content, and I don't expect it to ever trigger prematurely for any reasonably sized text, or even for a reasonably sized image.
Now if - on an empty page - you explicitly set `self.y` to a position lower than `self.t_margin` before inserting something huge, I'd say it is your own job to check first that it actually fits (or just disable auto breaking to begin with).

The "problem" being "solved" here really can only manifest itself in an already pathological situation, which probably doesn't justify the hassle. I'm not so much worried about an extra instance attribute, but maintaining this one in a reliable manner appears to be a rather messy task. It is already tricky to implement correctly, and will be highly error prone to work with in the context of a subclass or other ways to extend the library. And however we try to do it, it will probably fail when a user starts moving back and forth between pages, which is a perfectly valid thing to do!

Given all that, my vote would be to shelve this idea for the moment. Given the way how fpdf2 is currently structured, making automatic page breaks absolutely fool proof is not a realistic undertaking. And this implementation may cause as many new problems as it solves existing ones.

Comment (Lucas-C) [reply fpdf/fpdf.py:3175]: I agree with you @gmischler 

I introduced the `self.y > self.t_margin` check in #535

I'm going to wait for @semaeostomea answer/comments now.
Let's discuss if we should do more about this or just accept it as an edge case that end users should deal with on their own

Comment (gmischler) [reply fpdf/fpdf.py:3175]: Btw.: Unfortunately, checking for those magic 11 bytes above doesn't reliably work either.  This test will give the wrong result when the user has changed some settings like the font or a color. Many of those changes get immediately written to the output stream for the current page, so that its contents will then be longer even though nothing visible has been added to the page.

Fpdf2 is currently a very open system, which is really a feature, as it gives it a lot of flexibility and makes it straightforward to use. At the same time some automated functions are offered, but they make the system at best "semi automated". It is hard to put guard rails around stuff, because the content streams in a PDF file are just an unstructured sequence of text instructions. There are places where we could probably add more structured levels of abstraction and assert a more tight control over individual settings (for text content some of that has already happened), but its not an easy task. 

So maybe we could say that this proposal comes a bit too early in the evolution of the library. Sorry for that... :wink:

Comment (gmischler) [reply fpdf/fpdf.py:3177]: > I also intended for oversized images to get cut off on all sides and aligned them to the center.

If you already know that your content won't fit the page in the first place and should be cropped, then the correct way to handle that (short of actually cropping it) is to turn off auto-breaking and create the pages new manually.

Given the current architecture of the library, I don't see a way to make this idea fool-proof and not end up with a maintenance nightmare. There are just too many ways a user can circumvent the automation features (including direct access to `._out()`, even if that may not be specifically encouraged).

Thanks for trying anyway, though!

Comment (semaeostomea) [reply fpdf/fpdf.py:3177]: @Lucas-C just to double-check since you neither told me to go ahead nor anyone closed the PR:
should I go ahead with my idea of splitting up the page content or is this idea shelved for the moment and the PR can be closed?

Comment (Lucas-C) [reply fpdf/fpdf.py:3177]: > should I go ahead with my idea of splitting up the page content or is this idea shelved for the moment and the PR can be closed?

I honestly would prefer <ins>not</ins>to go this road for now, so I'm going to close this PR.
Thank you very much @semaeostomea for contributing this suggestion!

**Comment by Lucas-C** [comment]
@allcontributors please add @semaeostomea for code
